### PR TITLE
Tmb check

### PR DESCRIPTION
**Pull request by bbolker**
This adds a test to see if the current installed version of TMB is more recent than the one that glmmTMB was built with. (The test could perhaps be back-ported to `TMB`, it is a generalization of the tests that TMB uses to determine whether a newer version of `Matrix` has been installed.)

As far as comments go, I'd be most interested in someone checking [the new help page](https://github.com/glmmTMB/glmmTMB/blob/TMB_check/glmmTMB/man/reinstalling.Rd) (`?reinstalling`/) and see if it seems accurate and useful.

I haven't actually tried a 'live' test (install older version of TMB; re-install glmmTMB; install newer version of TMB and make sure that the warning is appropriately thrown) yet ...

### CODE REVIEW

**Comment by bbolker** [comment]
Not sure why this is failing with a TMB mismatch check??

**Comment by kaskr** [comment]
```
glmmTMB was built with TMB version 1.7.19
Current TMB version is 1.7.21
```

It's using an old cached glmmTMB build - why? The whole idea of the CI is to rebuild the package every time...

**Comment by kaskr** [comment]
I think you checked the TMB package version into the repository - was that intended?

File: glmmTMB/inst/TMB-version [hunk 1]
```diff
@@ -0,0 +1 @@
+1.7.19
```
Comment (kaskr) [inline glmmTMB/inst/TMB-version:1]: I don't think this should be in the repository.

File: glmmTMB/R/utils.R [hunk 2]
```diff
@@ -213,9 +213,69 @@ nullSparseMatrix <- function() {
         do.call(Matrix::sparseMatrix, c(argList, list(repr="T")))
     }
 }
+    
+## Check for version mismatch in dependent binary packages
+#' @importFrom utils packageVersion
+checkDepPackageVersion <- function(dep_pkg="TMB",this_pkg="glmmTMB",write_file=FALSE) {
+    cur_dep_version <- as.character(packageVersion(dep_pkg))
+    fn <- sprintf("%s-version",dep_pkg)
+    if (write_file) {
```
Comment (kaskr) [inline glmmTMB/R/utils.R:222]: I think the file should be written at install time. It seems to me it never is (`write_file=FALSE`)?

**Comment by bbolker** [comment]
I think you're probably right about all of this. I'll need to get back and check/see what's going on.  It should be stored at the time the package is **built** (which is not necessarily when it is installed, e.g. if someone installs from a binary then this should be the version of TMB/Matrix/etc. that the glmmTMB .so files were compiled against ...)

**Comment by kaskr** [comment]
Yes, I meant to say **built**

**Comment by bbolker** [comment]
Have implemented this in what *might* be the right way (see `src/Makevars` and the appropriate part of *Writing R Extensions* (search for "phony")), but I'm not going to stay up to see if it passes checks or not.

**Comment by bbolker** [comment]
OK, I think this is ready to merge ...

**Comment by kaskr** [comment]
Not a big deal, but personally I had preferred to get the warning *once* when doing `library(glmmTMB)` rather than 6 times after running  `example(glmmTMB)`...
I also don't like the fact that every model fitting now requires a disk access (again, not a big deal - had just expected something closer to zero overhead):

```
> system.time(glmmTMB:::checkDepPackageVersion())
   user  system elapsed 
  0.000   0.002   0.008
```

In any case it would be easy to fix by moving the version check somewhere else - I'll leave it to you to decide.

**Comment by bbolker** [comment]
That makes sense, I'll see what I can do.

**Comment by bbolker** [comment]
Seems easy/sensible to move the `checkDepPackageVersion()` call to `.onLoad` (in `zzz.R`) ?

**Comment by kaskr** [comment]
Right, that would be my preference too - as in https://github.com/kaskr/adcomp/blob/master/TMB/R/zzz.R

**Pull request merged**